File: ai-studyhub-frontend/src/services/adminService.js
import { apiGet, apiPost, apiPut, apiDelete } from './api'

/**
 * Ban user account
 * Bans a user account by setting their status to BANNED.
 * Method: POST
 * Path: /api/admin/users/{id}/ban
 * @param {any} id - Path parameter
 */
export function banUser(id) {
  return apiPost(`/admin/users/${id}/ban`);
}

/**
 * Create document category
 * Adds a new category name.
 * Method: POST
 * Path: /api/admin/categories
 * @param {any} body - Request body
 */
export function createCategory(body) {
  return apiPost(`/admin/categories`, body);
}

/**
 * Create course config
 * Adds a new course configuration linked to a Major.
 * Method: POST
 * Path: /api/admin/courses
 * @param {any} body - Request body
 */
export function createCourse(body) {
  return apiPost(`/admin/courses`, body);
}

/**
 * Create major config
 * Adds a new major config to the system.
 * Method: POST
 * Path: /api/admin/majors
 * @param {any} body - Request body
 */
export function createMajor(body) {
  return apiPost(`/admin/majors`, body);
}

/**
 * Create plan config
 * Adds a new subscription plan with specific limits and price.
 * Method: POST
 * Path: /api/admin/plans
 * @param {any} body - Request body
 */
export function createPlan(body) {
  return apiPost(`/admin/plans`, body);
}

/**
 * Delete document category
 * Deletes a document category by ID.
 * Method: DELETE
 * Path: /api/admin/categories/{id}
 * @param {any} id - Path parameter
 */
export function deleteCategory(id) {
  return apiDelete(`/admin/categories/${id}`);
}

/**
 * Delete course config
 * Deletes a course config by ID.
 * Method: DELETE
 * Path: /api/admin/courses/{id}
 * @param {any} id - Path parameter
 */
export function deleteCourse(id) {
  return apiDelete(`/admin/courses/${id}`);
}

/**
 * Delete major config
 * Deletes major configuration from the system.
 * Method: DELETE
 * Path: /api/admin/majors/{id}
 * @param {any} id - Path parameter
 */
export function deleteMajor(id) {
  return apiDelete(`/admin/majors/${id}`);
}

/**
 * Delete plan config
 * Deletes plan configuration by ID.
 * Method: DELETE
 * Path: /api/admin/plans/{id}
 * @param {any} id - Path parameter
 */
export function deletePlan(id) {
  return apiDelete(`/admin/plans/${id}`);
}

/**
 * List all document categories
 * Retrieves all document categories.
 * Method: GET
 * Path: /api/admin/categories
 */
export function getAllCategories() {
  return apiGet(`/admin/categories`);
}

/**
 * List all courses config
 * Retrieves all course configurations in the system.
 * Method: GET
 * Path: /api/admin/courses
 */
export function getAllCourses() {
  return apiGet(`/admin/courses`);
}

/**
 * Get all documents
 * Retrieves all documents uploaded in the system for administrative audit.
 * Method: GET
 * Path: /api/admin/documents
 */
export function getAllDocuments() {
  return apiGet(`/admin/documents`);
}

/**
 * List all majors config
 * Retrieves all major configurations in the system.
 * Method: GET
 * Path: /api/admin/majors
 */
export function getAllMajors() {
  return apiGet(`/admin/majors`);
}

/**
 * List all plans config
 * Retrieves all subscription plans configs in the system.
 * Method: GET
 * Path: /api/admin/plans
 */
export function getAllPlans() {
  return apiGet(`/admin/plans`);
}

/**
 * Get all reports
 * Retrieves all content violation reports submitted by users.
 * Method: GET
 * Path: /api/admin/reports
 */
export function getAllReports() {
  return apiGet(`/admin/reports`);
}

/**
 * Get all users list
 * Retrieves a list of all users registered in the system.
 * Method: GET
 * Path: /api/admin/users
 */
export function getAllUsers() {
  return apiGet(`/admin/users`);
}

/**
 * Get pending verifications
 * Retrieves all pending student verifications for review.
 * Method: GET
 * Path: /api/admin/verifications/pending
 */
export function getPendingVerifications() {
  return apiGet(`/admin/verifications/pending`);
}

/**
 * Moderate uploaded document
 * Approves or rejects a document to control visibility in the community feed.
 * Method: POST
 * Path: /api/admin/documents/{id}/moderate
 * @param {any} id - Path parameter
 * @param {any} body - Request body
 */
export function moderateDocument(id, body) {
  return apiPost(`/admin/documents/${id}/moderate`, body);
}

/**
 * Resolve violation report
 * Resolves a user report and optionally rejects/bans the reported document.
 * Method: POST
 * Path: /api/admin/reports/{id}/resolve
 * @param {any} id - Path parameter
 * @param {any} body - Request body
 */
export function resolveReport(id, body) {
  return apiPost(`/admin/reports/${id}/resolve`, body);
}

/**
 * Review student identity verification card
 * Approves or rejects a student identity card verification request.
 * Method: POST
 * Path: /api/admin/verifications/{id}/review
 * @param {any} id - Path parameter
 * @param {any} body - Request body
 */
export function reviewVerification(id, body) {
  return apiPost(`/admin/verifications/${id}/review`, body);
}

/**
 * Unban user account
 * Unbans a user account by restoring their status to ACTIVE.
 * Method: POST
 * Path: /api/admin/users/{id}/unban
 * @param {any} id - Path parameter
 */
export function unbanUser(id) {
  return apiPost(`/admin/users/${id}/unban`);
}

/**
 * Update document category
 * Updates category name by ID.
 * Method: PUT
 * Path: /api/admin/categories/{id}
 * @param {any} id - Path parameter
 * @param {any} body - Request body
 */
export function updateCategory(id, body) {
  return apiPut(`/admin/categories/${id}`, body);
}

/**
 * Update course config
 * Updates course properties by ID.
 * Method: PUT
 * Path: /api/admin/courses/{id}
 * @param {any} id - Path parameter
 * @param {any} body - Request body
 */
export function updateCourse(id, body) {
  return apiPut(`/admin/courses/${id}`, body);
}

/**
 * Update major config
 * Updates major properties by ID.
 * Method: PUT
 * Path: /api/admin/majors/{id}
 * @param {any} id - Path parameter
 * @param {any} body - Request body
 */
export function updateMajor(id, body) {
  return apiPut(`/admin/majors/${id}`, body);
}

/**
 * Update plan config
 * Updates subscription plan limits and prices by ID.
 * Method: PUT
 * Path: /api/admin/plans/{id}
 * @param {any} id - Path parameter
 * @param {any} body - Request body
 */
export function updatePlan(id, body) {
  return apiPut(`/admin/plans/${id}`, body);
}

File: ai-studyhub-backend/src/main/java/com/studyhub/admin/service/AdminService.java
package com.studyhub.admin.service;

import com.studyhub.admin.dto.*;
import com.studyhub.common.PageResponse;
import com.studyhub.common.enums.ModerationStatus;
import com.studyhub.common.enums.NotificationType;
import com.studyhub.common.enums.ReportStatus;
import com.studyhub.common.enums.UserStatus;
import com.studyhub.course.dto.CourseListResponse;
import com.studyhub.course.entity.Course;
import com.studyhub.course.entity.Major;
import com.studyhub.course.service.CourseService;
import com.studyhub.document.entity.Document;
import com.studyhub.document.entity.DocumentCategory;
import com.studyhub.document.entity.Report;
import com.studyhub.document.repository.DocumentRepository;
import com.studyhub.document.repository.ReportRepository;
import com.studyhub.document.service.DocumentService;
import com.studyhub.user.entity.SubscriptionPlan;
import com.studyhub.user.entity.User;
import com.studyhub.user.repository.UserRepository;
import com.studyhub.user.service.NotificationService;
import com.studyhub.user.service.SubscriptionService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.core.io.ByteArrayResource;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

@Slf4j
@Service
@RequiredArgsConstructor
public class AdminService {

    private final UserRepository userRepository;
    private final DocumentRepository documentRepository;
    private final ReportRepository reportRepository;
    private final NotificationService notificationService;

    // Decomposed Services
    private final CourseService courseService;
    private final SubscriptionService subscriptionService;
    private final DocumentService documentService;
    private final AdminVerificationService adminVerificationService;
    private final AdminAnalyticsService adminAnalyticsService;

    // ── 1. Quản lý người dùng (User Management) ─────────────────────────────────

    @Transactional(readOnly = true)
    public List<AdminUserResponse> getAllUsers() {
        log.info("Admin: Fetching all users list");
        return userRepository.findAll().stream()
                .sorted(Comparator.comparing(User::getCreatedAt, Comparator.nullsLast(Comparator.naturalOrder())).reversed())
                .map(this::mapUserResponse)
                .collect(Collectors.toList());
    }

    @Transactional
    public void banUser(Long userId) {
        log.info("Admin: Banning user ID {}", userId);
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new IllegalArgumentException("User not found"));
        user.setStatus(UserStatus.BANNED);
        userRepository.save(user);
    }

    @Transactional
    public void unbanUser(Long userId) {
        log.info("Admin: Unbanning user ID {}", userId);
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new IllegalArgumentException("User not found"));
        user.setStatus(UserStatus.ACTIVE);
        userRepository.save(user);
    }

    // ── 2. Analytics & Activity Logs (Delegated) ───────────────────────────────

    @Transactional(readOnly = true)
    public AdminDashboardAnalyticsResponse getDashboardAnalytics() {
        return adminAnalyticsService.getDashboardAnalytics();
    }

    @Transactional(readOnly = true)
    public PageResponse<AdminActivityLogResponse> getActivityLogs(
            String query,
            String type,
            LocalDate dateFrom,
            LocalDate dateTo,
            int page,
            int size
    ) {
        return adminAnalyticsService.getActivityLogs(query, type, dateFrom, dateTo, page, size);
    }

    @Transactional(readOnly = true)
    public ResponseEntity<ByteArrayResource> exportActivityLogs(
            String query,
            String type,
            LocalDate dateFrom,
            LocalDate dateTo
    ) {
        return adminAnalyticsService.exportActivityLogs(query, type, dateFrom, dateTo);
    }

    // ── 3. Duyệt yêu cầu xác thực thẻ sinh viên (Delegated) ─────────────────────

    @Transactional(readOnly = true)
    public List<AdminVerificationResponse> getPendingVerifications() {
        return adminVerificationService.getPendingVerifications();
    }

    @Transactional
    public void reviewVerification(Long verificationId, VerificationReviewRequest request, String adminEmail) {
        adminVerificationService.reviewVerification(verificationId, request, adminEmail);
    }

    // ── 4. Duyệt/Moderate tài liệu & Reports ────────────────────────────────────

    @Transactional(readOnly = true)
    public List<AdminDocumentResponse> getAllDocuments() {
        log.info("Admin: Fetching all documents");
        return documentRepository.findAll().stream()
                .map(doc -> AdminDocumentResponse.builder()
                        .id(doc.getId())
                        .title(doc.getTitle())
                        .fileName(doc.getFileName())
                        .fileSize(doc.getFileSize())
                        .visibility(doc.getVisibility() != null ? doc.getVisibility().name() : null)
                        .moderationStatus(doc.getModerationStatus() != null ? doc.getModerationStatus().name() : null)
                        .ownerEmail(doc.getUser().getEmail())
                        .averageRating(doc.getAverageRating())
                        .createdAt(doc.getCreatedAt())
                        .build())
                .collect(Collectors.toList());
    }

    @Transactional
    public void moderateDocument(Long documentId, DocumentModerationRequest request) {
        log.info("Admin: Moderating document ID {} status to {}", documentId, request.getStatus());
        Document document = documentRepository.findById(documentId)
                .orElseThrow(() -> new IllegalArgumentException("Document not found"));

        ModerationStatus targetStatus = ModerationStatus.valueOf(request.getStatus().toUpperCase());
        document.setModerationStatus(targetStatus);
        documentRepository.save(document);

        String documentTitle = document.getTitle() != null && !document.getTitle().isBlank()
                ? document.getTitle()
                : "your document";

        if (targetStatus == ModerationStatus.APPROVED) {
            notificationService.createNotification(
                    document.getUser(),
                    "Document Approved",
                    "Your document \"" + documentTitle + "\" has been approved and published.",
                    NotificationType.DOCUMENT
            );
        } else if (targetStatus == ModerationStatus.REJECTED) {
            String reason = request.getReason() != null && !request.getReason().isBlank()
                    ? " Reason: " + request.getReason().trim()
                    : "";
            notificationService.createNotification(
                    document.getUser(),
                    "Document Rejected",
                    "Your document \"" + documentTitle + "\" was rejected." + reason,
                    NotificationType.DOCUMENT
            );
        }
    }

    @Transactional(readOnly = true)
    public List<AdminReportResponse> getAllReports() {
        log.info("Admin: Fetching all user reports");
        return reportRepository.findAll().stream()
                .map(report -> AdminReportResponse.builder()
                        .id(report.getId())
                        .reporterEmail(report.getReporter().getEmail())
                        .documentId(report.getDocument().getId())
                        .documentTitle(report.getDocument().getTitle())
                        .reportType(report.getReportType().name())
                        .status(report.getStatus().name())
                        .reportReason(report.getReportReason())
                        .createdAt(report.getCreatedAt())
                        .build())
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public AdminReportDetailResponse getReportDetail(Long reportId) {
        log.info("Admin: Fetching detail for report ID {}", reportId);
        Report report = reportRepository.findById(reportId)
                .orElseThrow(() -> new IllegalArgumentException("Report not found"));

        Document document = report.getDocument();
        User reporter = report.getReporter();

        return AdminReportDetailResponse.builder()
                .id(report.getId())
                .reporterEmail(reporter != null ? reporter.getEmail() : null)
                .reporterFullName(reporter != null ? reporter.getFullName() : null)
                .documentOwnerEmail(document != null && document.getUser() != null ? document.getUser().getEmail() : null)
                .documentId(document != null ? document.getId() : null)
                .documentTitle(document != null ? document.getTitle() : null)
                .documentFileName(document != null ? document.getFileName() : null)
                .documentVisibility(document != null && document.getVisibility() != null ? document.getVisibility().name() : null)
                .documentModerationStatus(document != null && document.getModerationStatus() != null ? document.getModerationStatus().name() : null)
                .courseCode(document != null && document.getCourse() != null ? document.getCourse().getCourseCode() : null)
                .courseName(document != null && document.getCourse() != null ? document.getCourse().getCourseName() : null)
                .documentReportCount(document != null ? reportRepository.countByDocumentId(document.getId()) : 0)
                .reportType(report.getReportType() != null ? report.getReportType().name() : null)
                .status(report.getStatus() != null ? report.getStatus().name() : null)
                .reportReason(report.getReportReason())
                .createdAt(report.getCreatedAt())
                .documentCreatedAt(document != null ? document.getCreatedAt() : null)
                .build();
    }

    @Transactional
    public void resolveReport(Long reportId, ReportResolveRequest request) {
        log.info("Admin: Resolving report ID to status {}", reportId, request.getStatus());
        Report report = reportRepository.findById(reportId)
                .orElseThrow(() -> new IllegalArgumentException("Report not found"));

        ReportStatus targetStatus = ReportStatus.valueOf(request.getStatus().toUpperCase());
        report.setStatus(targetStatus);
        reportRepository.save(report);

        // Xoá (từ chối duyệt) tài liệu nếu yêu cầu
        if (request.isDeleteDocument() && targetStatus == ReportStatus.RESOLVED) {
            Document doc = report.getDocument();
            doc.setModerationStatus(ModerationStatus.REJECTED);
            documentRepository.save(doc);
        }
    }

    // ── 5. Quản lý cấu hình hệ thống (Delegated) ───────────────────────────────

    // --- Majors CRUD
    @Transactional(readOnly = true)
    public List<Major> getAllMajors() {
        return courseService.getAllMajors();
    }

    @Transactional
    public Major createMajor(MajorRequest request) {
        return courseService.createMajor(request);
    }

    @Transactional
    public Major updateMajor(Long id, MajorRequest request) {
        return courseService.updateMajor(id, request);
    }

    @Transactional
    public void deleteMajor(Long id) {
        courseService.deleteMajor(id);
    }

    // --- Courses CRUD
    @Transactional(readOnly = true)
    public List<CourseListResponse> getAllCourses() {
        return courseService.getAllCourses();
    }

    @Transactional
    public CourseListResponse createCourse(CourseRequest request) {
        return courseService.createCourse(request);
    }

    @Transactional
    public CourseListResponse updateCourse(Long id, CourseRequest request) {
        return courseService.updateCourse(id, request);
    }

    @Transactional
    public void deleteCourse(Long id) {
        courseService.deleteCourse(id);
    }

    // --- Categories CRUD
    @Transactional(readOnly = true)
    public List<DocumentCategory> getAllCategories() {
        return documentService.getAllCategories();
    }

    @Transactional
    public DocumentCategory createCategory(CategoryRequest request) {
        return documentService.createCategory(request);
    }

    @Transactional
    public DocumentCategory updateCategory(Long id, CategoryRequest request) {
        return documentService.updateCategory(id, request);
    }

    @Transactional
    public void deleteCategory(Long id) {
        documentService.deleteCategory(id);
    }

    // --- Subscription Plans CRUD
    @Transactional(readOnly = true)
    public List<SubscriptionPlan> getAllPlans() {
        return subscriptionService.getAllPlans();
    }

    @Transactional
    public SubscriptionPlan createPlan(SubscriptionPlanRequest request) {
        return subscriptionService.createPlan(request);
    }

    @Transactional
    public SubscriptionPlan updatePlan(Long id, SubscriptionPlanRequest request) {
        return subscriptionService.updatePlan(id, request);
    }

    @Transactional
    public void deletePlan(Long id) {
        subscriptionService.deletePlan(id);
    }

    @Transactional(readOnly = true)
    public List<com.studyhub.user.dto.SubscriptionPlanVersionResponse> getPlanVersions(Long id) {
        return subscriptionService.getPlanVersions(id);
    }

    private AdminUserResponse mapUserResponse(User user) {
        return AdminUserResponse.builder()
                .id(user.getId())
                .fullName(user.getFullName())
                .email(user.getEmail())
                .verificationStatus(user.getVerificationStatus() != null ? user.getVerificationStatus().name() : null)
                .status(user.getStatus() != null ? user.getStatus().name() : null)
                .planName(user.getPlan() != null ? user.getPlan().getPlanName() : "FREE")
                .roleName(user.getRole() != null ? user.getRole().getRoleName() : "USER")
                .createdAt(user.getCreatedAt())
                .build();
    }
}
